### ISSUE TITLE
Invalid recursive methods or functions

### ISSUE THREAD

**Issue by ghosh-rounak** [microsoft/TypeScript#55643]
### 🔍 Search Terms

Invalid recursive methods or functions not detected as error by typescript compiler during compile time

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

// TypeScript compiler should detect invalid recursive functions and methods

// Invalid Recursions should not compile and should be reported as error during compile time

// Examples :
function testFunction(i:number) {
  return testFunction(i); // compiler should warn this as invalid recursive function
}

let testFunction2 = function (i:number) {
  return testFunction2(i); // compiler should warn this as invalid recursive function
}

let testFunction3 :(i:number)=>number = (i:number)=>{
  return testFunction3(i); // compiler should warn this as invalid recursive function
}

testFunction(1); // compiler should warn this as invalid recursive function call
testFunction2(1); // compiler should warn this as invalid recursive function call
testFunction3(1); // compiler should warn this as invalid recursive function call

class Testing {
  id:number=0;

  test1():number{
    return this.test1(); // compiler should warn this as invalid recursive method
  }
}

const testing:Testing = new Testing();
console.log(testing.test1());// compiler should warn this as invalid recursive method call

//-----------------------------------------------------------------------------------------

// Correct Recursions should work fine and should not be reported as error by compiler 

// Examples :
function factorial(x:number):number {

  // if number is 0
  if (x == 0) {
      return 1;
  }

  // if number is positive
  else {
      return x * factorial(x - 1);
  }
}

console.log(factorial(5));

### 📃 Motivating Example

// Invalid Recursions should not compile and should be reported as error during compile time

// Examples :
function testFunction(i:number) {
  return testFunction(i); // compiler should warn this as invalid recursive function
}

let testFunction2 = function (i:number) {
  return testFunction2(i); // compiler should warn this as invalid recursive function
}

let testFunction3 :(i:number)=>number = (i:number)=>{
  return testFunction3(i); // compiler should warn this as invalid recursive function
}

testFunction(1); // compiler should warn this as invalid recursive function call
testFunction2(1); // compiler should warn this as invalid recursive function call
testFunction3(1); // compiler should warn this as invalid recursive function call

class Testing {
  id:number=0;

  test1():number{
    return this.test1(); // compiler should warn this as invalid recursive method
  }
}

const testing:Testing = new Testing();
console.log(testing.test1());// compiler should warn this as invalid recursive method call

//-----------------------------------------------------------------------------------------

// Correct Recursions should work fine and should not be reported as error by compiler 

// Examples :
function factorial(x:number):number {

  // if number is 0
  if (x == 0) {
      return 1;
  }

  // if number is positive
  else {
      return x * factorial(x - 1);
  }
}

console.log(factorial(5));

### 💻 Use Cases

// Invalid Recursions should not compile and should be reported as error during compile time

// Examples :
function testFunction(i:number) {
  return testFunction(i); // compiler should warn this as invalid recursive function
}

let testFunction2 = function (i:number) {
  return testFunction2(i); // compiler should warn this as invalid recursive function
}

let testFunction3 :(i:number)=>number = (i:number)=>{
  return testFunction3(i); // compiler should warn this as invalid recursive function
}

testFunction(1); // compiler should warn this as invalid recursive function call
testFunction2(1); // compiler should warn this as invalid recursive function call
testFunction3(1); // compiler should warn this as invalid recursive function call

class Testing {
  id:number=0;

  test1():number{
    return this.test1(); // compiler should warn this as invalid recursive method
  }
}

const testing:Testing = new Testing();
console.log(testing.test1());// compiler should warn this as invalid recursive method call

//-----------------------------------------------------------------------------------------

// Correct Recursions should work fine and should not be reported as error by compiler 

// Examples :
function factorial(x:number):number {

  // if number is 0
  if (x == 0) {
      return 1;
  }

  // if number is positive
  else {
      return x * factorial(x - 1);
  }
}

console.log(factorial(5));

1. What do you want to use this for?
Java programmers can detect invalid recursive methods and functions in Intellij.
Swift programmers can detect invalid recursive methods and functions in Xcode.
But typescript programmers cannot detect invalid recursive methods and functions in VsCode or any other code editor.
Invalid recursion check should exist in typescript as well which will allow to catch bugs during compile time .

3. What shortcomings exist with current approaches?
Java programmers can detect invalid recursive methods and functions in Intellij.
Swift programmers can detect invalid recursive methods and functions in Xcode.
But typescript programmers cannot detect invalid recursive methods and functions in VsCode or any other code editor.
Invalid recursion check should exist in typescript as well which will allow to catch bugs during compile time .

5. What workarounds are you using in the meantime?
No workarounds at the moment for detecting typescript invalid recursive methods or functions during compile time .
As a result runtime error happens for invalid recursive methods or functions in typescript.

**Comment by MartinJohns** [comment]
You really don't need to repeat the example code in every section of the template. This makes your issues incredible hard to read. I would suggest you to edit your issues to make them readable.

**Comment by ghosh-rounak** [comment]
Can a flag be introduced in tsconfig.json to detect invalid recursive functions and methods like the one mentioned below ?

`function testFunction( i : number ) {
return testFunction(i); // compiler should warn this as invalid recursive function during compile time 
}`

**Comment by xiBread** [comment]
TypeScript is not a linter.

**Comment by ghosh-rounak** [comment]
> TypeScript is not a linter.
Is there any eslint rule or any other vscode extention using which we can detect  invalid recursive functions and methods in TypeScript?

In Java we get warning in Intellij community edition for invalid recursive functions and methods. Can we get similar warning in VsCode while using TypeScript?

**Comment by ghosh-rounak** [comment]
> Can you please put your code in code blocks? Like this
> 
> ```
> ```ts
> code
> ```
> ```
> 
> Also if you're expecting TypeScript to behave like Java, you're just not going to have a good time in general.

```
function testFunction( i : number ) : number{
/*
compiler should warn the below line as invalid
recursive function during compile time
*/
return testFunction(i);
}

testFunction(5);
```

Is it possible to detect the above invalid recursive function as error or warning during compile time by typescript ?
